### PR TITLE
Dev br

### PR DESCRIPTION
**Pull request by ivddorrka**

### CODE REVIEW

File: ixwebsocket/proxysocteksetting.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#ifndef proxysocteksettings_h
+#define proxysocteksettings_h
```
Comment (stark-dev) [inline ixwebsocket/proxysocteksetting.h]: fix typo

File: ixwebsocket/IXProxySocketSetting.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#include "IXProxySocketSetting.h"
+
+#include <iostream>
+
+void ProxySetup::setProxyPass(std::string &proxypass) {
```
Comment (noiseonwires) [inline ixwebsocket/IXProxySocketSetting.cpp]: If you are using a reference (and not modifying it), always use const reference.

If you are taking an argument and saving it into a variable (class member), better use pass-by-value + std::move

Same for below

File: ixwebsocket/IXProxySocketSetting.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    connection_web = 0x20,
+    connection_sock4 = 0x04,
+    connection_sock5 = 0x05
+
+};
+
+struct ProxySetup {
+
+public:
+
```
Comment (noiseonwires) [inline ixwebsocket/IXProxySocketSetting.h]: redundant spaces

File: ixwebsocket/IXSocketConnect.cpp [hunk 4]
```diff
@@ -24,6 +25,10 @@
 
 namespace ix
 {
+    void logger(__attribute__((unused)) int level, const char* message, __attribute__((unused)) void* userdata)
```
Comment (noiseonwires) [inline ixwebsocket/IXSocketConnect.cpp]: better prefer `[[maybe unused]]`, as it's a part of the C++ standard, and __attribute__ is a GCC extension.

Or even simpler way: use just a type (`int`) without argument name, and the compiler will not complain.

File: ixwebsocket/IXSocketConnect.cpp [hunk 5]
```diff
@@ -148,4 +160,34 @@ namespace ix
...
+                                                    proxy_settings.get_proxy_pass().c_str());
+        proxysocketconfig_set_logging(proxyConfig, logger, nullptr);
+        int fd = proxysocket_connect(proxyConfig, host.c_str(), port, &errorMsg);
+
+        if (fd == -1)
+        {
+            errMsg = errorMsg;
+            return -1;
+        }
+        proxysocketconfig_free(proxyConfig);
```
Comment (noiseonwires) [inline ixwebsocket/IXSocketConnect.cpp]: If the previous condition is true (`fd == -1`), then you'll get a memory leak because `free` will be not called.
Use std::unique_ptr.

**Review by noiseonwires — CHANGES_REQUESTED** [review]

File: main.cpp [hunk 6]
```diff
@@ -14,21 +14,37 @@
...
+    ProxySetup setup_params;
+    ProxyConnectionType proxy_con_type_choice = ProxyConnectionType::connection_web;
+
+    std::string proxytype("WEB");
+    std::string proxyhost("proxy.emea.etn.com");
+    setup_params.setProxyHost(proxyhost);
+    setup_params.setProxyPort(8080);
+    setup_params.setProxyType(proxy_con_type_choice);
+
+    webSocket.setProxySettings(std::ref(setup_params));
```
Comment (noiseonwires) [inline main.cpp]: No need to use std::ref here.

File: main.cpp [hunk 7]
```diff
@@ -14,21 +14,37 @@
  */
 
 #include <ixwebsocket/IXNetSystem.h>
```
Comment (stark-dev) [inline main.cpp]: We can skip these changes in the patch file, as the main file is not used in our websocket wrapper

File: ixwebsocket/IXProxySocketSetting.cpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
+#include "IXProxySocketSetting.h"
+
+#include <iostream>
+
+void ix::ProxySetup::setProxyPass(std::string &proxypass) {
```
Comment (noiseonwires) [inline ixwebsocket/IXProxySocketSetting.cpp]: Still not right :)
Let me explain it in more details.

in C++ there are two main types of values: rvalues and lvalues (yes, it means "left" and "right" generally).
lvalues are, for example, just variables and class members; rvalues are, for example, some temporary objects (like the function returns something and it's immediately used).

Now let's imagine, that you have a method and inside it you want to save the value to the class member.
If you have `setProxyPass(const std::string& proxypass)  { proxyPass = proxypass; }`
then it will be:
if setProxyPass() called with lvalue as an argument: 1 copy (inside the function)
if setProxyPass() called with rvalue as an argument: 1 copy (inside the function)

If you have `setProxyPass(std::string proxypass)  { proxyPass = std::move(proxypass); }`
then it will be
if setProxyPass() called with lvalue as an argument: 1 copy (before the call)
if setProxyPass() called with rvalue as an argument: 0 copies! (temporary object is moved directly to the class member)

As std::string objects with strings longer than some length (you can google for SSO - short string optimization) are kept in heap and there's no CoW (copy-on-write) in modern C++ standards, usually when the object is copied it causes new memory allocation, and that's a very slow operation (x100-x1000 times slower than just copy some registers or stack-allocated data).

The approach that you have currently in code is the worst one - you pass the argument as a non-const reference and do std::move for it. So for rvalues it's OK, but for lvalues it will implicitly clean the variable on the caller side that may lead to bugs and undefined behavior.

TL;DR
1. If you are using references, always use const references until there's a big need to modify it in-place
2. If you accept an argument and save it to the class member, use pass-by-value (not reference!) + std::move
